source/m2mtlvserializer.cpp

```diff
@@ -152,87 +155,76 @@ void M2MTLVSerializer::serialize_resource_instance(uint16_t id, M2MResourceInsta
 void M2MTLVSerializer::serialize_TILV(uint8_t type, uint16_t id, uint8_t *value, uint32_t value_length, uint8_t *&data, uint32_t &size)
 {
     uint8_t *tlv = 0;
-    uint32_t type_length = 1;
+    const uint32_t type_length = TLV_TYPE_SIZE;
     type += id < 256 ? 0 : ID16;
     type += value_length < 8 ? value_length :
             value_length < 256 ? LENGTH8 :
             value_length < 65536 ? LENGTH16 : LENGTH24;
-    uint8_t *tlv_type = (uint8_t*)malloc(type_length+1);
-    memset(tlv_type,0,type_length+1);
-    *tlv_type = type & 0xFF;
+    uint8_t tlv_type;
+    tlv_type = type & 0xFF;
 
-    uint32_t id_size = 0;
-    uint8_t* id_ptr = serialize_id(id, id_size);
+    uint32_t id_size;
+    uint8_t id_array[MAX_TLV_ID_SIZE];
+    serialize_id(id, id_size, id_array);
 
-    uint32_t length_size = 0;
-    uint8_t* length_ptr = serialize_length(value_length, length_size);
+    uint32_t length_size;
+    uint8_t length_array[MAX_TLV_LENGTH_SIZE];
+    serialize_length(value_length, length_size, length_array);
 
-    tlv = (uint8_t*)malloc(size + type_length + id_size + length_size + value_length+1);
-    memset(tlv,0,size + type_length + id_size + length_size + value_length+1);
+    tlv = (uint8_t*)malloc(size + type_length + id_size + length_size + value_length);
     if(data) {
         memcpy(tlv, data, size);
+        free(data);
+    }
+
+    uint8_t *out_ptr = tlv + size;
+    *out_ptr++=tlv_type;
+
+    int i=0;
+    while(i<id_size) {
+        *out_ptr++=id_array[i++];
     }
-    memcpy(tlv+size, tlv_type, type_length);
-    memcpy(tlv+size+type_length, id_ptr, id_size);
-    memcpy(tlv+size+type_length+id_size, length_ptr, length_size);
-    memcpy(tlv+size+type_length+id_size+length_size, value, value_length);
 
```

Comment:
I'm not fan of hand-coding the memcpy(). If you use the memcpy(), a sane compiler will select the suitable implementation based on compilation strategy chosen by developer. eg. if size optimization is chosen, the compiler will emit a call to memcpy(), but unroll that if speed optimization was chosen. In any case, the code like this is not performance critical and smallest code footprint is usually what we want.